olive/common/ort_inference.py

```diff
@@ -51,26 +75,299 @@ def get_ort_inference_session(
+            raise ValueError("'provider_options' should be a sequence.")
+
+        if len(providers) != len(provider_options):
+            raise ValueError("'providers' and 'provider_options' should be the same length if both are given.")
+
+        if not all(isinstance(provider, str) for provider in providers):
+            raise ValueError("Only string values for 'providers' are supported if 'provider_options' is given.")
+
+        if not all(isinstance(options_for_provider, dict) for options_for_provider in provider_options):
+            raise ValueError("'provider_options' values must be dicts.")
+
+        for name, options in zip(providers, provider_options):
+            set_provider_options(name, options)
+    else:
+        for provider in providers:
+            if isinstance(provider, str):
+                set_provider_options(provider, {})
+            elif (
+                isinstance(provider, (tuple, list))
+                and len(provider) == 2
+                and isinstance(provider[0], str)
+                and isinstance(provider[1], dict)
+            ):
+                set_provider_options(provider[0], provider[1])
+            else:
+                raise ValueError("'providers' values must be either strings or (string, dict) tuples.")
+
+    return list(provider_name_to_options.keys()), list(provider_name_to_options.values())
+
+
+def check_ort_fallback(session: "InferenceSession", execution_providers: Sequence[str]):
+    """Check if the onnxruntime fallback happens and raise an error if it does."""
+    # pylint: disable=protected-access
+    session_providers = session.get_providers()
+    for ep in execution_providers:
+        if ep not in session_providers:
+            raise OrtSessionFallbackError(
+                f"The onnxruntime fallback happens. {ep} is not in the session providers {session_providers}."
+                f" session._enable_fallback = {session._enable_fallback}"
+            )
+    session.disable_fallback()
+
+
+class OrtInferenceSession:
+    """ORT Inference Session with IO binding."""
+
+    def __init__(
+        self,
+        session: "InferenceSession",
+        io_bind: bool = False,
+        device: str = "cpu",
+        shared_kv_buffer: bool = False,
+        use_fp16: bool = False,
+        input_feed: Optional[Dict[str, np.ndarray]] = None,
+    ):
+        self.session = session
+        self.io_bind = io_bind
+        self.device = device
+        self.shared_kv_buffer = shared_kv_buffer
+        self.use_fp16 = use_fp16
+        self.kv_cache_ortvalues = {} if (self.shared_kv_buffer and self.use_fp16) else None
+
+        self.io_binding = None
+        if self.io_bind:
+            self.io_binding = self.session.io_binding()
+            if self.shared_kv_buffer and self.use_fp16:
+                assert input_feed is not None, "input_feed is required when shared_kv_buffer and use_fp16 are True"
+                bind_input_data(
```

Comment:
Missed the "device_id"?

Reply:
I didn't add it here since the original code I abstracted didn't use device_id. but I can add it here for completeness if you think it's better.